src/server/firefox.ts

```diff
@@ -108,39 +107,36 @@ export class Firefox extends BrowserTypeBase {
   }
 }
 
-function wrapTransportWithWebSocket(transport: ConnectionTransport, logger: InnerLogger, port: number): WebSocketWrapper {
-  const server = new ws.Server({ port });
-  const guid = helper.guid();
-  const idMixer = new SequenceNumberMixer<{id: number, socket: ws}>();
+type SessionData = {
+ socket: ws,
+};
+
+function startWebSocketServer(transport: ConnectionTransport, logger: InnerLogger, port: number): WebSocketServer {
   const pendingBrowserContextCreations = new Set<number>();
   const pendingBrowserContextDeletions = new Map<number, string>();
   const browserContextIds = new Map<string, ws>();
-  const sessionToSocket = new Map<string, ws>();
-  const sockets = new Set<ws>();
+  const sessionToData = new Map<string, SessionData>();
+
+  function removeSession(sessionId: string): SessionData | undefined {
+    const data = sessionToData.get(sessionId);
+    if (!data)
+      return;
+    sessionToData.delete(sessionId);
+    return data;
+  }
 
-  transport.onmessage = message => {
-    if (typeof message.id === 'number') {
+  const server = new WebSocketServer(transport, logger, port, {
+    onBrowserResponse(seqNum: number, source: ws, message: ProtocolResponse) {
       // Process command response.
-      const seqNum = message.id;
-      const value = idMixer.take(seqNum);
-      if (!value)
-        return;
-      const { id, socket } = value;
-
-      if (socket.readyState === ws.CLOSING) {
-        if (pendingBrowserContextCreations.has(id)) {
-          transport.send({
-            id: ++SequenceNumberMixer._lastSequenceNumber,
-            method: 'Browser.removeBrowserContext',
-            params: { browserContextId: message.result.browserContextId }
-          });
-        }
+      if (source.readyState === ws.CLOSING) {
```

Comment:
I think we do not handle the case when client detaches after pendingBrowserContextCreations has been sent, but before its response has been received.

Reply:
Confused - you are pointing to the line where we `Browser.removeBrowserContext`, what else would you like to happen?

Reply:
Figured it out